File: programs/java/techniques/recursion/intermediate/TowerOfHanoi.java
package programs.java.techniques.recursion.intermediate;

public class TowerOfHanoi {

    /**
     * Move 'n' disks from the Source rod to the Destination rod using the Helper
     * rod.
     *
     * Rules:
     * 1. Only one disk can be moved at a time.
     * 2. A larger disk cannot be placed on top of a smaller disk.
     * 3. Only the top disk of a rod can be moved.
     *
     * Recursive Idea:
     *
     * To move 'n' disks from Source -> Destination:
     *
     * Step 1:
     * Move the top (n - 1) disks from Source to Helper using Destination as the
     * temporary rod.
     *
     * Step 2:
     * Move the largest (nth) disk from Source to Destination.
     *
     * Step 3:
     * Move the (n - 1) disks from Helper to Destination using Source as the
     * temporary rod.
     *
     * Base Case:
     * If only one disk is left, move it directly from Source to Destination.
     *
     * Time Complexity:
     * T(n) = 2T(n - 1) + 1
     * => O(2^n)
     *
     * Auxiliary Space:
     * O(n)
     * (Maximum recursion depth is n.)
     *
     * @param n           Number of disks to move.
     * @param source      The rod from which the disks are moved.
     * @param helper      The temporary rod used during the movement.
     * @param destination The rod to which the disks are moved.
     */
    public static void move(int n, char source, char helper, char destination) {
        // Base Case:
        // If there is only one disk, move it directly.
        if (n == 1) {
            System.out.println("Move 1 from " + source + " to " + destination);
            return;
        }

        // Step 1:
        // Move the top (n - 1) disks from Source to Helper.
        move(n - 1, source, destination, helper);

        // Step 2:
        // Move the largest disk from Source to Destination.
        System.out.println("Move " + n + " from " + source + " to " + destination);

        // Step 3:
        // Move the (n - 1) disks from Helper to Destination.
        move(n - 1, helper, source, destination);
    }

    public static void main(String[] args) {
        move(3, 'A', 'B', 'C');
    }
}

File: programs/cpp/techniques/recursion/tower_of_hanoi.cpp
#include <iostream>

using namespace std;

/**
 * Prints the steps required to move n disks from the source rod
 * to the destination rod using the auxiliary rod.
 *
 * Approach:
 * - To move n disks, first move n - 1 disks from source to auxiliary.
 * - Move the largest disk (n) from source to destination.
 * - Move the remaining n - 1 disks from auxiliary to destination.
 * - The same process is recursively repeated for n - 1 disks.
 *
 * Base case:
 * - If n == 0, there are no disks to move, so return.
 *
 * Time Complexity: O(2^n)
 * Space Complexity: O(n) - recursion stack
 */
void move_disks(int n, char source, char auxiliary, char destination)
{
    // Base case: no disks to move.
    if (n == 0)
        return;

    // Move n - 1 disks from source to auxiliary
    // using destination as temporary storage.
    move_disks(n - 1, source, destination, auxiliary);

    // Move the largest disk from source to destination.
    cout << "Move " << n << " from " << source << " to " << destination << endl;

    // Move n - 1 disks from auxiliary to destination
    // using source as temporary storage.
    move_disks(n - 1, auxiliary, source, destination);
}

/**
 * Starts the Tower of Hanoi process using the default rods:
 * A as source, B as auxiliary, and C as destination.
 */
void move_disks(int n)
{
    move_disks(n, 'A', 'B', 'C');
}

int main()
{
    int disks = 3;

    cout << "Steps to move " << disks << " disks: " << endl;

    move_disks(disks);

    return 0;
}
